File: Remove and Reverse - GFG/remove-and-reverse.py
#{ 
 # Driver Code Starts
#Initial Template for Python 3

# } Driver Code Ends
#User function Template for python3
class Solution:
    def removeReverse(self, S): 
        dic={}
        n=len(S)
        s=list(S)
        for i in S:
            if i in dic:
                dic[i]+=1
            else:
                dic[i]=1
                
        i=0
        j=n-1
        dir=1
        while(i<=j):
            if dir==1:
                if dic[S[i]]==1:
                    i+=1
                else:
                    dic[S[i]]-=1
                    s[i]='@'
                    i+=1
                    dir=-1
            else:
                if dic[S[j]]==1:
                    j-=1
                else:
                    dic[S[j]]-=1
                    s[j]='@'
                    j-=1
                    dir=1
                
        
        ans=""
        for i in s:
            if i!='@':
                ans=ans+i
        
       
        if dir==-1:
            return ans[::-1]
        
        return ans
        
        
        
        

#{ 
 # Driver Code Starts.
if __name__ == '__main__': 
    t = int(input())
    for _ in range(t):
        S = input()
        ob = Solution()
        ans = ob.removeReverse(S)
        print(ans)


# } Driver Code Ends

File: Remove and Reverse - GFG/remove-and-reverse.cpp
//{ Driver Code Starts
//Initial Template for C++
#include <bits/stdc++.h>
using namespace std;

// } Driver Code Ends
//User function Template for C++
class Solution {
  public:
    string removeReverse(string s) {
        vector<int> vec(26,0); 
        for(auto it:s){
            vec[it-'a']++;
          }
        int dir=1;
        int i=0, j=s.size()-1;
        while(i<=j){
            if(dir==1){
                if(vec[s[i]-'a']==1) i++;
                else{
                    vec[s[i]-'a']--;
                    s[i]='@';
                    i++;
                    dir=-1;
                }   
            }
            
            else{
                if(vec[s[j]-'a']==1) j--;
                else{
                    vec[s[j]-'a']--;
                    s[j]='@';
                    j--;
                    dir=1;
                }  
            }
        }
        
        string ans="";
        for(auto it:s){
            if(it!='@') ans.push_back(it); 
        }
        if(dir==-1) reverse(ans.begin(),ans.end());
        return ans;
        
    }
};

//{ Driver Code Starts.
int main() {
    int t;
    cin >> t;
    while (t--) {
        string S;
        cin>>S;
        Solution obj;
        auto ans = obj.removeReverse(S);
        cout<<ans<<endl;
    }
}
// } Driver Code Ends